synedrion/src/cggmp21/sigma/elog.rs

```diff
@@ -0,0 +1,180 @@
+//! Dlog with El-Gamal Commitment ($\Pi^{elog}$, Section A.1, Fig. 23)
+
+use core::marker::PhantomData;
+
+use rand_core::CryptoRngCore;
+use serde::{Deserialize, Serialize};
+
+use super::super::SchemeParams;
+use crate::{
+    curve::{Point, Scalar},
+    tools::{
+        hashing::{Chain, Hashable, XofHasher},
+        Secret,
+    },
+};
+
+const HASH_TAG: &[u8] = b"P_elog";
+
+pub(crate) struct ElogSecretInputs<'a> {
+    pub y: &'a Secret<Scalar>,
+    pub lambda: &'a Secret<Scalar>,
+}
+
+pub(crate) struct ElogPublicInputs<'a> {
+    /// Point $L = g * \lambda$, where $g$ is the curve generator.
+    pub cap_l: &'a Point,
+    /// Point $M = g * y + X * \lambda$, where $g$ is the curve generator.
+    pub cap_m: &'a Point,
+    /// Point $X$, satisfying the condition above.
+    pub cap_x: &'a Point,
+    /// Point $Y = h * y$.
+    pub cap_y: &'a Point,
+    /// Point $h$, satisfying the condition above.
+    pub h: &'a Point,
+}
+
+/// ZK proof: Paillier decryption modulo $q$.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub(crate) struct ElogProof<P: SchemeParams> {
+    e: Scalar,
+    cap_a: Point,
+    cap_n: Point,
+    cap_b: Point,
+    z: Scalar,
+    u: Scalar,
+    phantom: PhantomData<P>,
+}
+
+impl<P: SchemeParams> ElogProof<P> {
+    pub fn new(
+        rng: &mut impl CryptoRngCore,
+        secret: ElogSecretInputs<'_>,
+        public: ElogPublicInputs<'_>,
+        aux: &impl Hashable,
+    ) -> Self {
+        let alpha = Secret::init_with(|| Scalar::random(rng));
+        let m = Secret::init_with(|| Scalar::random(rng));
+
+        let cap_a = alpha.mul_by_generator();
+        let cap_n = m.mul_by_generator() + public.cap_x * &alpha;
+        let cap_b = public.h * &m;
+
+        let mut reader = XofHasher::new_with_dst(HASH_TAG)
+            // commitments
+            .chain(&cap_a)
+            .chain(&cap_n)
+            .chain(&cap_b)
+            // public parameters
+            .chain(&public.cap_l)
+            .chain(&public.cap_m)
+            .chain(&public.cap_x)
+            .chain(&public.cap_y)
+            .chain(&public.h)
+            .chain(aux)
+            .finalize_to_reader();
+
+        // Non-interactive challenge
+        let e = Scalar::from_xof_reader(&mut reader);
+
+        let z = *(alpha + secret.lambda * e).expose_secret();
+        let u = *(m + secret.y * e).expose_secret();
```

Comment:
Same question here I guess: alpha, lambda and e are in the order of the curve, so z and u are 512 bit right? Where is the `mod q` part? I mean, given q is 1024-bit, it's trivially true that z and z mod q is the same value, but is that all there is to it?

Reply:
> Where is the mod q part?

`alpha`, `lambda`, and `e` are all of `Scalar` type, so `mod q` happens automatically on any arithmetic operations with them.

Reply:
Since I wrote this question I think my confusion stems from thinking that `q` is one of the two primes for an RSA modulus, but this is a different `q` isn't it? It's the prime of the curve prime field yes?